### PR TITLE
multiple traces with new decimator api (#1062)

### LINKED ISSUE #1062: BUG - Decimator does not work with multiple traces
## Without decimator:
image

## With decimator:
```bash
[2024-03-28 15:10:31][Taipy][INFO]  * Server starting on http://127.0.0.1:5000
C:\Users\asaju\Desktop\report_issues\venv\Lib\site-packages\taipy\gui\gui.py:635: TaipyGuiWarning: A problem occurred while resolving variable 'data' in module '__main__'.
  _warn(
```

image

## Code:
```python
from taipy.gui import Gui
from taipy.gui.data.decimator import MinMaxDecimator
import pandas as pd
import numpy as np

NOP = 500
decimator_instance = MinMaxDecimator(n_out=NOP)


data = pd.DataFrame(
    {
        "Time": np.arange(0, 10, 0.1),
        "A": np.arange(0, 10, 0.1) + np.random.randn(100) + 10,
        "B": np.arange(0, 10, 0.1) + np.random.randn(100),
    }
)

page = """
<|{data}|chart|x=Time|y[1]=A|y[2]=B|decimator=decimator_instance|>
"""

Gui(page).run()
```

### LINKED ISSUE #1515: Possibility to use decimator on Plotly charts
### Description

The goal would be to allow users to use the decimator alongside Plotly charts.

```python
import yfinance as yf
from taipy.gui import Gui
from taipy.gui.data.decimator import MinMaxDecimator, RDP, LTTB


df_AAPL = yf.Ticker("AAPL").history(interval="1d", period = "max")
df_AAPL["DATE"] = df_AAPL.index.astype('int64').astype(float)

n_out = 50
decimator_instance = MinMaxDecimator(n_out=n_out)

decimate_data_count = len(df_AAPL)

import plotly.express as px

fig = px.line(df_AAPL, x="DATE", y="Open", markers=True)

page = """
# Decimator

From a data length of <|{len(df_AAPL)}|> to <|{n_out}|>

## Without decimator

<|chart|figure={fig}|>

## With decimator

<|chart|figure={fig}|decimator=decimator_instance|>
"""

gui = Gui(page)
gui.run(port=5026, title="Decimator")

```

### Acceptance Criteria

- [ ] Ensure new code is unit tested, and check code coverage is at least 90%.
- [ ] Propagate any change on the demos and run all of them to ensure there is no breaking change.
- [ ] Ensure any change is well documented.

### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

### PR DESCRIPTION
**Pull request by dinhlongviolin1**
Resolve #1062

Resolve #1515

This new api will also enable the use of applying the decimator directly to an existing dataframe -> use for plotly charts

### CODE REVIEW

File: frontend/taipy-gui/src/components/Taipy/Chart.tsx [hunk 1]
```diff
@@ -151,21 +151,26 @@ const getDecimatorsPayload = (
...
-                            chartMode: modes[i],
-                        }
-                      : undefined
-              ),
-              relayoutData: relayoutData,
-          }
+            width: plotDiv?.clientWidth,
+            height: plotDiv?.clientHeight,
+            decimators: decimators.map((d, i) =>
+                d
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/src/components/Taipy/Chart.tsx]: Wouldn't it be more readable this way?
```suggestion
                desc = {
                        xAxis: getAxis(traces, i, columns, 0),
                        yAxis: getAxis(traces, i, columns, 1),
                        zAxis: getAxis(traces, i, columns, 2),
                        chartMode: modes[i],
                }
                if (d) {
                    desc["decimator"] = d;
                }
                return desc
```

File: taipy/gui/data/decimator/base.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+                return self.__user_defined_apply_decimator(
+                    dataframe, x_column_name, y_column_name, z_column_name, payload, is_copied
+                )
+            except Exception as e:
+                _warn("Error executing user defined apply_decimator function: ", e)
+        return self._df_apply_decimator(dataframe, x_column_name, y_column_name, z_column_name, payload, is_copied)
+
+    @abstractmethod
+    def decimate(self, data: np.ndarray, payload: t.Dict[str, t.Any]) -> np.ndarray:
+        """Decimate function.
```
Comment (FabienLelaquais) [inline taipy/gui/data/decimator/base.py]:
```suggestion
        """NOT DOCUMENTED
        Decimate the dataset.
```

**Comment by FlorianJacta** [comment]
Is this normal that this was closed?

File: taipy/gui/data/decimator/__init__.py [hunk 3]
```diff
@@ -9,7 +9,11 @@
 # an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 # specific language governing permissions and limitations under the License.
 
+from .base import Decimator
 from .lttb import LTTB
 from .minmax import MinMaxDecimator
 from .rdp import RDP
 from .scatter_decimator import ScatterDecimator
+
+# Export the following namespace
```
Comment (FabienLelaquais) [inline taipy/gui/data/decimator/__init__.py]:
```suggestion
# Export the following classes
```

File: taipy/gui/data/decimator/base.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+    different use cases.
+    """
+
+    _CHART_MODES: t.List[str] = []
+
+    def __init__(
+        self,
+        threshold: t.Optional[int],
+        zoom: t.Optional[bool],
+        on_decimate: t.Optional[t.Callable] = None,
```
Comment (FabienLelaquais) [inline taipy/gui/data/decimator/base.py]: Please comment this out.

File: taipy/gui/data/decimator/base.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+        apply_decimator: t.Optional[t.Callable] = None,
```
Comment (FabienLelaquais) [inline taipy/gui/data/decimator/base.py]: Please comment this out - for the time being.

File: taipy/gui/data/decimator/base.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+    ) -> None:  # noqa: E501
+        """Initialize a new `Decimator`.
+
+        Arguments:
+            threshold (Optional[int]): The minimum amount of data points before the
+                decimator class is applied.
+            zoom (Optional[bool]): set to True to reapply the decimation
+                when zoom or re-layout events are triggered.
+            on_decimate (Optional[Callable]): an user-defined function that is executed when the decimator
```
Comment (FabienLelaquais) [inline taipy/gui/data/decimator/base.py]: Please remove the doc for these parameters - place them in a comment elsewhere so the doc doesn't mess up.

File: taipy/gui/data/decimator/base.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+
+    def on_decimate(
+        self,
+        df: pd.DataFrame,
+        decimator_instance_payload: t.Dict[str, t.Any],
+        decimator_payload: t.Dict[str, t.Any],
+        is_copied: bool = False,
+        filter_unused_columns: bool = True,
+    ):
+        """This function is executed whenever a decimator is found during runtime.
```
Comment (FabienLelaquais) [inline taipy/gui/data/decimator/base.py]:
```suggestion
        """NOT DOCUMENTED
        
        This function is executed whenever a decimator is found during runtime.
```

File: taipy/gui/data/decimator/base.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+    def apply_decimator(
+        self,
+        dataframe: pd.DataFrame,
+        x_column_name: t.Optional[str],
+        y_column_name: str,
+        z_column_name: str,
+        payload: t.Dict[str, t.Any],
+        is_copied: bool,
+    ):
+        """This function is executed whenever a decimator is applied to the data.
```
Comment (FabienLelaquais) [inline taipy/gui/data/decimator/base.py]:
```suggestion
        """NOT DOCUMENTED
        This function is executed whenever a decimator is applied to the data.
```

File: taipy/gui/data/decimator/lttb.py [hunk 9]
```diff
@@ -28,7 +28,14 @@ class LTTB(Decimator):
 
     _CHART_MODES = ["lines+markers", "lines", "markers"]
 
-    def __init__(self, n_out: int, threshold: t.Optional[int] = None, zoom: t.Optional[bool] = True) -> None:
+    def __init__(
+        self,
+        n_out: int,
+        threshold: t.Optional[int] = None,
+        zoom: t.Optional[bool] = True,
+        on_decimate: t.Optional[t.Callable] = None,
```
Comment (FabienLelaquais) [inline taipy/gui/data/decimator/lttb.py]: Please comment these two parameters out.

File: taipy/gui/data/decimator/lttb.py [hunk 10]
```diff
@@ -37,8 +44,12 @@ def __init__(self, n_out: int, threshold: t.Optional[int] = None, zoom: t.Option
                 decimation is applied.
             zoom (Optional[bool]): set to True to reapply the decimation
                 when zoom or re-layout events are triggered.
+            on_decimate (Optional[Callable]): an user-defined function that is executed when the decimator
```
Comment (FabienLelaquais) [inline taipy/gui/data/decimator/lttb.py]: Please move the doc for the two parameters outside this comment block.

File: taipy/gui/data/decimator/minmax.py [hunk 11]
```diff
@@ -36,8 +43,12 @@ def __init__(self, n_out: int, threshold: t.Optional[int] = None, zoom: t.Option
                 decimation is applied.
             zoom (Optional[bool]): set to True to reapply the decimation
                 when zoom or re-layout events are triggered.
+            on_decimate (Optional[Callable]): an user-defined function that is executed when the decimator
```
Comment (FabienLelaquais) [inline taipy/gui/data/decimator/minmax.py]: Please move the doc for the two parameters outside this comment block.

File: taipy/gui/data/decimator/rdp.py [hunk 12]
```diff
@@ -48,8 +50,13 @@ def __init__(
                 decimation is applied.
             zoom (Optional[bool]): set to True to reapply the decimation
                 when zoom or re-layout events are triggered.
+            on_decimate (Optional[Callable]): an user-defined function that is executed when the decimator
```
Comment (FabienLelaquais) [inline taipy/gui/data/decimator/rdp.py]: Please move the doc for the two parameters outside this comment block.

File: taipy/gui/data/decimator/minmax.py [hunk 13]
```diff
@@ -27,7 +27,14 @@ class MinMaxDecimator(Decimator):
 
     _CHART_MODES = ["lines+markers", "lines", "markers"]
 
-    def __init__(self, n_out: int, threshold: t.Optional[int] = None, zoom: t.Optional[bool] = True):
+    def __init__(
+        self,
+        n_out: int,
+        threshold: t.Optional[int] = None,
+        zoom: t.Optional[bool] = True,
+        on_decimate: t.Optional[t.Callable] = None,
```
Comment (FabienLelaquais) [inline taipy/gui/data/decimator/minmax.py]: Please comment these two parameters out.

File: taipy/gui/data/decimator/rdp.py [hunk 14]
```diff
@@ -34,6 +34,8 @@ def __init__(
         n_out: t.Optional[int] = None,
         threshold: t.Optional[int] = None,
         zoom: t.Optional[bool] = True,
+        on_decimate: t.Optional[t.Callable] = None,
```
Comment (FabienLelaquais) [inline taipy/gui/data/decimator/rdp.py]: Please comment these two parameters out.

File: taipy/gui/data/decimator/scatter_decimator.py [hunk 15]
```diff
@@ -34,6 +34,8 @@ def __init__(
         max_overlap_points: t.Optional[int] = None,
         threshold: t.Optional[int] = None,
         zoom: t.Optional[bool] = True,
+        on_decimate: t.Optional[t.Callable] = None,
```
Comment (FabienLelaquais) [inline taipy/gui/data/decimator/scatter_decimator.py]: Please comment these two parameters out.

File: taipy/gui/data/decimator/scatter_decimator.py [hunk 16]
```diff
@@ -47,6 +49,10 @@ def __init__(
                 decimation is applied.
             zoom (Optional[bool]): set to True to reapply the decimation
                 when zoom or re-layout events are triggered.
+            on_decimate (Optional[Callable]): an user-defined function that is executed when the decimator
```
Comment (FabienLelaquais) [inline taipy/gui/data/decimator/scatter_decimator.py]: Please move the doc for the two parameters outside this comment block.

**Review by FabienLelaquais — CHANGES_REQUESTED** [review]
Please remove the two new parameters for Decimator and inherited classes.
That will be brought back the day after 4.0 is out.

**Review by FabienLelaquais — DISMISSED** [review]
👍

**Review by FabienLelaquais — APPROVED** [review]
👍

**Pull request merged**